### ISSUE TITLE
Arbitrary subplot layouts

### ISSUE THREAD

**Issue by tbreloff** [tbreloff/Plots.jl#60]
Right now I support a "grid layout" and a "flex layout".  The latter is actually limited in that I can stretch a plot horizontally but not vertically.  i.e. I can do this:

```
----------------------
|          |         |
----------------------
|                    |
----------------------
```

but not this:

```
----------------------
|          |         |
-----------|         |
|          |         |
----------------------
```

I'd like to be able to support arbitrary spans, and likely allow passing in of matrices and recursive structures of Plot objects for creation.

**Comment by tbreloff** [comment]
Quick note: it should be possible to specify subplot coordinates exactly (relative to the full figure).  So I'd like to be able to do:

```
subplot(rand(10,2), layout = [((0,1), (0,1)), ((0.7,1), (0,0.3))])
```

And this should create an inset plot where the first plot takes up the full area, and the 2nd plot is inset in the top-right corner.

**Comment by tbreloff** [comment]
>  allow passing in of matrices and recursive structures of Plot objects for creation.

@spencerlyon2 I guess this has been sitting on the back-burner for a while.  Right now, I'm thinking about the design.

Doing this:
```julia
y = rand(10,3)
subplot(y, layout = @layout [[0.2,_] [0.3]])
```

could create something like:

```
----------------------
|    1     |         |
-----------|    3    |
|    2     |         |
----------------------
```

where `1` takes up 20% of the height, and `3` takes up 30% of the width.

I want the layout mechanic to be:

- concise code
- recursive structure
- ability to specify sizes (in absolute and percentage terms)

Lets brainstorm on the best API, and we can figure out the best way to implement it later.

**Comment by timholy** [comment]
There were some old mailing list discussions about this, but for some reason I can't find them. I wrote [this package](https://github.com/timholy/Layout.jl) (long outdated) as an attempt to make progress. Since then, I believe @dcjones added a linear programming-based layout manager to Gadfly, but I haven't really looked.

**Comment by tbreloff** [comment]
Thanks for the link @timholy.  It looks like yours is more ambitious than what I was considering (I was envisioning a manual layout... not "automated to optimize aesthetics")... but I'll see if the ideas overlap well.

As for Gadfly... do you mean this? https://github.com/dcjones/Gadfly.jl/blob/fcb249691700cdfde56184049060258ca8758c6d/src/geom/subplot.jl#L139

I'll be honest that I find the Gadfly code hard to follow, so I can't be sure that I fully understand what that code is doing.

**Comment by tbreloff** [comment]
In this example, the "outer array" is a 1x2 matrix (a row-vector), so it represents a "row".  For each item in the row, it could be a plot, or it could be another (nested) layout.

> Does the _ mean "fill the rest of the space in this column"?

I think it should mean "fill in the rest of the available space equally".  So if it's `[0.2, _, _]`, there would be 3 plots in a column, the first plot would take up 20% of the height and the other two would take up 40% each.

```
----------------------
|          |         |
----------------------
|                    |
----------------------
```

I would do:  `[[_ _], _]`, replacing the underscores with percentages if I want uneven sizes.

**Comment by spencerlyon2** [comment]
Ahh I see. Ok I think this could be great. It seems to be an extension of the idea I implemented in PlotlyJS (extension in that is supports nesting).

I'll continue to try to poke holes in the proposed API and see if we can make it more effective.

**Comment by timholy** [comment]
I can't find an optimizing layout manager in Gadfly, so I must be mistaken.

I think a string macro to parse those ASCII art samples and convert them into layout would be awesome :smile:.

**Comment by tbreloff** [comment]
I was thinking the same thing @timholy... it might be a little too much typing though.

**Comment by pkofod** [comment]
However, if you use `unicodeplots()` it would really be in the spirit of the backend to do it like that!

**Comment by spencerlyon2** [comment]
I could pass the display of a unicode plot to the layout macro ;)

**Comment by tbreloff** [comment]
Maybe:

```julia
p1,p2,p3 = [plot(rand(10)) for i=1:3]
subplot(layout"""
p1 | p2
----------
p3
"""
```

**Comment by pkofod** [comment]
Maybe just 
```
p1,p2,p3 = [plot(rand(10)) for i=1:3]
subplot(layout"""
p1 | p2
---
p3
"""
```
or would you just allow an arbitrary number of `-`'s? 10 `-`'s seems excessive ;)

**Comment by tbreloff** [comment]
I'm working through whether this type will cover what we need in a nice way:

```julia
immutable FlexLayout <: SubplotLayout
    n::Int
    grid::Matrix{Union{Int,FlexLayout}} # Nested layouts. Each position
                            # can be a plot index or another FlexLayout
    widths::Vector{Float64}
    heights::Vector{Float64}
end
```

Any standard grid is covered by it, though we can allow arbitrary nesting for any layout.  I'll work through the implementation, but we should continue to brainstorm on the cleanest and most convenient interface(s) for building this object.

**Comment by tbreloff** [comment]
@spencerlyon2 In the plotly backend (not plotlyjs) I set up subplots as independent html divs, bypassing Plotly's subplotting mechanism all together.  I do something similar for Gadfly/Immerse.

I'm wondering if I can continue to get away with that... I suppose there's no reason I can't stack a bunch of PlotlyJS divs into the same cell output and control them independently, right?

I guess the downside is when you'd like to link/share axes between plots.  I'm not sure if it's worth the tradeoff of ease/simplicity.

**Comment by timholy** [comment]
@tbreloff, that looks like a very nice foundation.

One question: in the grid you're laying out, do the "grid lines" align to the *outer bounds* of a plot, or to the *axis bounds*? By *outer bounds* I mean the full area corresponding to a single plot, which would include the x- and y- ticks and axis labels; *axis bounds* correspond to just the rectangular area in which the data points are represented.

The reason I ask is that it seems you're setting up a situation in which the interface specifies the outer bounds, which is very nice as it corresponds to how users would typically want to think about the problem. But in terms of the visual appeal of the final product, it will be weird if the axis bounds don't align on a grid, too. If there aren't good tools for achieving that, then one alternative is to let your grid representation specify the axis bounds, and have a separate notation indicating that particular rows/columns are left as blank space for labels.

Perhaps you could even use the same data structure, and allow a special index to indicate "blank." You could probably even automatically compute the "padded" version (axis bounds) from the user's input (outer bounds).

**Comment by tbreloff** [comment]
@timholy these are all really nice comments, but I'm not sure that Plots has this level of fine-tuning available to it. I guess the best you can really hope for is that you give rough heuristics (split the width 70/30, link the y-axes) and hope the backend has the toolset to account for labels/axes in an aesthetically pleasing way. I'm open to suggestions!

**Comment by tbreloff** [comment]
To expand a little on the problem... very frequently one needs access to a lot of low-level internal information in order to leave proper amounts of room for axis labels and ticks.  This depends on many things, including rendered font size, orientation, etc.  It's a pretty hard problem to solve this generically for all backends.  I do agree with your idea, though, that a 70/30 split should apply to the "plot grids", excluding the space for labels/guides.  Whether it is practically feasible to accomplish this with most backends is another story.  My experience is that backends typically give you the choice:

- I'll do the layout for you, or
- You tell me **exactly** where to place everything.

**Comment by tbreloff** [comment]
matplotlib ref: http://matplotlib.org/users/gridspec.html
maybe some good ideas: http://web.stanford.edu/~mwaskom/software/seaborn/tutorial/axis_grids.html
plotly: https://plot.ly/javascript/subplots/

In Gadfly, it may be possible to render the plots without displaying them, traverse the internals of the Context to figure out the sizes of the guides, then concatenate the contexts together with recalculated sizes.

**Comment by spencerlyon2** [comment]
@tbreloff I think that putting the plots in their own div and using html to align them is only a short-run solution. My main concerns are:

1. I can't save a figure (to svg, png,  pdf, etc.)  with subplots using plotly.js's export routines if each subplot is in it's div. They need to be in one div to get them exported properly.
2. Each subplot will get its own legend and menu bar -- not ideal aesthetically
3. When I do hook up javascript event support, users would have to register callbacks with each plot in the subplot rather than with the single plot that contains multiple subplots. 

So I guess the more precise version of what I'm trying to say is that eventually I want to support the arbitrary subplot layouts you've been talking about here, but do so by leveraging plotly's internal subplotting mechanisms.

**Comment by tbreloff** [comment]
I've been rebuilding some internals of Plots to make recipes first class citizens, and as part of that I (intentionally) broke subplots as they exist.  This issue is now at the forefront of my brainstorming, as I want to rebuild the subplot mechanic, hopefully integrating it into the recipe concept, so that I can dispatch on subplot layouts and avoid re-coding everything for both plots and subplots.

If you have any ideas for features for subplots, now's the time to share!

**Comment by tbreloff** [comment]
I've been thinking through this problem a little more thoroughly.  I think I've got a rough design sketch that will work.

Definitions:

- `Plot`: the whole "figure", data, attributes, layout... everything
- `Layout`: organization/positions of components
- `Axis`: holds limits, ticks, etc (can be shared!)
- `AxisView`: wraps a reference to the Axis, plus guides
- `PlotArea`: the actual plotting area (inside spines/borders) (may not need an actual type for this)
- `Subplot`: container for PlotArea, AxisViews, titles, etc
- `Series`: wraps a combined "data + attributes" for a single series (i.e. a path, or a histogram, etc) (could map to more than one Subplot!)

Rules:

- Subplots are no longer a separate type that encompass `Plot` objects... that definition is reversed (a `Plot` contains one or more `Subplot`)
- Subplots are first-class members... a single plot is really just a single `Subplot` taking up the full space of a `Plot`
- The list of series belong to the `Plot`, and each stores a reference to the `Subplot` that it belongs to
- A recursive grid structure `GridLayout` will determine the position of each subplot, as well as which subplots share alignments (for example, the top borders of the "plot area" should line up)
- An alternative layout could define precise coordinates (for inset plots)
- Layouts will contain settings such as margin, padding, etc
- Layouts will be defined by percentages of width/height relative to their parent layouts, and the default strategy will be to apply those percentages to the `PlotArea`, not the `Subplot` (addresses @timholy's concerns), and may require an iterative approach in calculating optimal guide sizes
- calling `plot(plt1, plt2, ...)` will build a new plot with subplots plt1, plt2... (there will no longer be a `subplot` method!)
- there will be a `subplot` keyword that will accept either a `Subplot` reference(s) or integer(s) representing the subplot index... this way you have full control over which series go to each subplot.  Passing a vector could map a series to more than one subplot.
- For convenience, if you set the subplot keyword but not the layout, it will build a layout for you.  If you set `subplot=true` it will build a layout for you, and also automatically assign series to the subplots in a round-robin fashion (similar to current behavior)

I'm still working through some of the design, but I think this is a good start.  Comments welcome!

**Comment by tbreloff** [comment]
I'm putting this here in case I forget about it.  For PyPlot, it will probably be easiest to [add the subplot axes to the figure directly](http://matplotlib.org/api/figure_api.html#matplotlib.figure.Figure.add_axes), and use `ax[:set_position](...)` to move them after creation.  Then I don't need to worry about hacking GridSpec/SubplotSpec and all the other matplotlib annoyances.

**Comment by tbreloff** [comment]
For anyone following along, I've made some great progress here (though there's still a bunch to do).    Notice how the plot areas line up nicely:

![tmp](https://cloud.githubusercontent.com/assets/933338/15359239/696be2a4-1cd6-11e6-8e5d-418756124bfe.png)

The implementation is pretty generic, and the pyplot-specific code is pretty minimal (just enough to compute some bounding boxes and later set the positions).  It should be reasonable to roll this out to other backends, and it should be "straightforward" to support all sorts of complex layouts, including inset/floating subplots, etc.

I'm using https://github.com/dcjones/Measures.jl to keep track of positions within the canvas... I wonder if something else might be better though.  I've seen https://github.com/ajkeller34/Unitful.jl which looks nice at first glance but might not support julia 0.4.  Is there anything else that might work for combining different coordinate units?  Should I stick with Measures?

Also are there any must-have features when laying out subplots?  Margin, padding, title positions, etc?

**Comment by lobingera** [comment]
looking at the right, upper subplot: Maybe something about propagating the size of the plot ... and all of the above.

**Comment by tbreloff** [comment]
> something about propagating the size of the plot 

Do you mean: defining the size (in percentage or fixed units) of one subplot in the grid, and then letting everything else fill in the available space?  If so, then yes that's on my list.

**Comment by diegozea** [comment]
Is will be a way to link the y axis scale/ticks row-wise and the x scale/ticks column-wise?I imagine: `fixedy=true` or something like `fixed=:none`, `fixed=:xy`, `fixed=:x` ...

**Comment by lobingera** [comment]
@tbreloff No, i meant, all the functions the plot a subplot need to be aware of the 'canvas' size (in absolute), so e.g. autolabeling, autoticks work correctly. Filling constraints, by adapting to the available size comes along that.

**Comment by tbreloff** [comment]
@diegozea Yes, and I envision this will be really nice.  The way I have it set up now... each subplot has several `Axis` objects corresponding to each axis, and which track the data being plotted to them.  Linking axes across subplots should be as easy as pointing both subplots to the same `Axis`.  Then limits, etc can be shared but you can control the view differently on each subplot.  Things like correlation grids and facets should be easy to make and style just as you want.

**Comment by tbreloff** [comment]
@lobingera I still think I'm confused what you mean.  Maybe you're referring to reducing the number of tick labels automatically so they don't overlap?  (Note: this is a great example of how matplotlib fails miserably when you let it choose things by itself)

**Comment by lobingera** [comment]
I think we mean the same: Every subplot needs to be aware of spacing and size, so that any auto scaling functionality can work correctly. 
Your comment on matplotlib failing miserably might be rephrased once you start to work on non-overlapping items in plots on your own...

**Comment by tbreloff** [comment]
> Your comment on matplotlib failing miserably might be rephrased once you start to work on non-overlapping items in plots on your own...

Maybe... we'll see.  Though matplotlib has been developed and used for over a decade.  You'd think they'd be able to avoid label overlap...

In this case, it seems simple enough to have a loop:

```julia
for i=1:length(ticklabels)-1
  t1, t2 = ticklabels[i:i+1]
  if overlaps(t1, t2)
    hide(t2)
  end
end
```

A real implementation would be nicer, probably adding a bounding box buffer so they're not too close together, but not too different I think.  This assumes you already have tick labels placed of course, though the logic for initial placement would be somewhat similar.

**Comment by tbreloff** [comment]
I'm thinking through nice ways to initialize layouts... it might be best done as as macro that operates on matrix initialization:

```
---------
| a |   |
|---| c |
| b |   |
---------
```
would be created by:
```julia
@layout [[a;b] c{0.25pct, 20mm}]
```

I'd just reach in to grab custom sizing from the `curly` Expr, otherwise just compose a bunch of `GridLayout` objects.  The symbols a/b/c could be used to retrieve the associated subplot (i.e. `plt[:a]` gives the top-left subplot)

Also:

- a tuple `(2,3)` would create a 2x3 grid of equal size subplots
- an integer would automatically choose the grid shape for that many subplots

**Comment by pkofod** [comment]
Looking good. I'm waiting for this to implement subplots in PGFPlots.. no pressure... ;)

(I would love more details as soon as you've converged on implementation details.)

**Comment by tbreloff** [comment]
Progress...

<img width="749" alt="screen shot 2016-05-18 at 11 49 22 pm" src="https://cloud.githubusercontent.com/assets/933338/15382011/352194d0-1d53-11e6-9da2-04d6445750ad.png">

The sizes are a little buggy, but getting closer.  Laying it out like this lets you do things like `plt[:c]` which will give you the bottom-left subplot.

**Comment by tbreloff** [comment]
I'm pretty happy with the layout mechanic now.  I'll update the docs soon with details of how it all works.  

![tmp](https://cloud.githubusercontent.com/assets/933338/15614133/0415c998-2405-11e6-9a1b-c7bffb73df98.png)

**Comment by gabrielgellner** [comment]
This looks amazing! The new 0.7 version makes me salivate for Julia plotting nirvana! Not sure if it is useful at this point, or at all, but I remember reading thru R's Grid graphics design ideas on layered `ViewPorts` and thought that maybe all the hard thinking they have done on these kinds of issues might help with any of the corner cases of the api that maybe you are still wondering about.

**Comment by tbreloff** [comment]
Thanks! Yeah I'm pretty excited about 0.7... Not sure there's much hard
thinking left to do!

**Comment by gabrielgellner** [comment]
Sweetness, that is what I want to hear!

 With all the sweet changes I think 0.7 is a little understated, should be 0.7! ~ 0.908639 ;)

Anyway back to playing with the dev branch :+1:
